pkg/core/blockchain.go

```diff
@@ -613,18 +624,231 @@ func (bc *Blockchain) jumpToStateInternal(p uint32, stage stateJumpStage) error
+	}
+	if height == currHeight && stage == none {
+		bc.log.Info("chain is already at the proper state", zap.Uint32("height", height))
+		return nil
+	}
+	if bc.config.KeepOnlyLatestState {
+		return fmt.Errorf("KeepOnlyLatestState is enabled, state for height %d is outdated and removed from the storage", height)
+	}
+	if bc.config.RemoveUntraceableBlocks && currHeight >= bc.config.MaxTraceableBlocks {
+		return fmt.Errorf("RemoveUntraceableBlocks is enabled, a necessary batch of traceable blocks has already been removed")
+	}
+
+	// Retrieve necessary state before the DB modification.
+	hHeight := bc.HeaderHeight()
+	b, err := bc.GetBlock(bc.headerHashes[height])
+	if err != nil {
+		return fmt.Errorf("failed to retrieve block %d: %w", height, err)
+	}
+	sr, err := bc.stateRoot.GetStateRoot(height)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve stateroot for height %d: %w", height, err)
+	}
+	v := bc.dao.Version
+	cache := bc.dao // dao is MemCachedStore over DB, so use dao directly to persist cached changes right to the underlying DB
+
+	bc.log.Info("initialize state reset", zap.Uint32("height", height))
+	start := time.Now()
+	p := start
+
+	resetStageKey := []byte{byte(storage.SYSStateChangeStage)}
+	switch stage {
+	case none:
+		cache.Store.Put(resetStageKey, []byte{byte(stateReset), byte(stateChangeStage1)})
+		fallthrough
+	case stateChangeStage1:
+		// Remove blocks/transactions/aers from currHeight down to height (not including height itself).
+		for i := height + 1; i <= currHeight; i++ {
+			err := cache.DeleteBlock(bc.headerHashes[i])
+			if err != nil {
+				return fmt.Errorf("error while removing block %d: %w", i, err)
+			}
+		}
+		for i := height + 1; i <= hHeight; i++ {
+			cache.DeleteHeader(bc.headerHashes[i])
+		}
+		cache.Store.Put(resetStageKey, []byte{byte(stateReset), byte(stateChangeStage2)})
+
+		bc.log.Info("blocks, transactions ans AERs are reset", zap.Duration("duration", time.Since(p)))
+		p = time.Now()
+		fallthrough
+	case stateChangeStage2:
+		// Completely remove contract IDs to update them later.
+		cache.Store.Seek(storage.SeekRange{Prefix: []byte{byte(storage.STContractID)}}, func(k, _ []byte) bool {
+			cache.Store.Delete(k)
+			return true
+		})
+
+		// Reset contracts storage and store new contract IDs.
+		var mode = mpt.ModeAll
+		if bc.config.RemoveUntraceableBlocks {
+			mode |= mpt.ModeGCFlag
+		}
+		sStore := storage.NewPrivateMemCachedStore(cache.Store)
+		trieStore := mpt.NewTrieStore(sr.Root, mode, sStore)
+		oldStoragePrefix := v.StoragePrefix
+		newStoragePrefix := statesync.TemporaryPrefix(oldStoragePrefix)
+		mgmtCSPrefixLen := 1 + 4 + 1 // STStorage + Management ID + contract state prefix
+		mgmtContractPrefix := make([]byte, mgmtCSPrefixLen-1)
+		id := int32(native.ManagementContractID)
+		binary.BigEndian.PutUint32(mgmtContractPrefix, uint32(id))
+		mgmtContractPrefix[4] = native.PrefixContract
+		cs := new(state.Contract)
+
+		// TODO: Hi, OOM killer
```

Comment:
Just flush it periodically. If it's to be restarted it's not a problem, the data would be the same.

Reply:
Then we have to remove items at the same Seek iteration, but it's OK. Fixed.